### PR TITLE
feat: Random dataset with specified input and output sequence length

### LINKED ISSUE #1302: Synthetic rollout length for GRPO performance benchmarking
**Is your feature request related to a problem? Please describe.**
Vllm benchmarking CLI supports ignoring EOS and synthesizing given OSL ([--random-output-len ](https://docs.vllm.ai/en/latest/cli/bench/serve.html?h=ignore_eos#-random-output-len)). Can we use this feature to generate synthetic benchmarks of fixed OSL for evaluating perf at certain sequence length scenario?

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by guyueh1**
# What does this PR do ?

Random dataset following specified input and output sequence length

# Issues
closes #1302 

# Usage
Use the following flags for fixed ISL/OSL eval
```bash
uv run examples/run_eval_random_dataset.py \
+data.input_len_or_input_len_generator=1000 \
generation.ignore_eos=true \
generation.vllm_cfg.max_model_len=3000
```

Use the following flags for fixed ISL/OSL GRPO
```bash
uv run examples/run_grpo_random_dataset.py \
+data.input_len_or_input_len_generator=1000 \
policy.generation.ignore_eos=true \
policy.generation.output_len_or_output_len_generator=2000 
```

Use the following flags for random ISL/OSL GRPO with mean + stdv
```bash
uv run examples/run_grpo_random_dataset.py \
grpo.val_at_start=false \
grpo.val_period=0 \
policy.max_total_sequence_length=8000 \
+data.input_len_or_input_len_generator.mean=1000 \
+data.input_len_or_input_len_generator.std=100 \
+policy.generation.output_len_or_output_len_generator.mean=2000 \
+policy.generation.output_len_or_output_len_generator.std=1000 \
policy.generation.ignore_eos=True
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

## Summary by CodeRabbit

* **New Features**
  * Added end-of-sequence handling configuration option across generation settings.
  * New example scripts for evaluation and training workflows with random math datasets.
  * Enhanced support for FP8 precision with Mixture-of-Experts models.
  * Added flexible output length configuration for generation tasks.

* **Chores**
  * Updated configuration files with new generation parameters.
  * Added diagnostic metrics output in training logs.

### CODE REVIEW

**Comment by guyueh1** [comment]
/ok to test [a934d95](https://github.com/NVIDIA-NeMo/RL/pull/1453/commits/a934d95972ce00f066733e6720aa20a78c19c464)

**Comment by guyueh1** [comment]
+ @yuki-97 to review since this is about dataset

File: examples/run_eval.py [hunk 1]
```diff
@@ -39,17 +49,54 @@ def parse_args():
...
-    args, remaining = parser.parse_known_args()
-
-    # Convert remaining args to OmegaConf format
-    overrides = OmegaConf.from_dotlist(remaining)
+    args, overrides = parser.parse_known_args()
 
     return args, overrides
 
 
+def _setup_random_data(tokenizer, data_config):
```
Comment (yuki-97) [inline examples/run_eval.py]: can we move this function (and the one in `run_grpo.py`) to `nemo_rl/data/utils.py` and reuse it?

File: nemo_rl/models/generation/vllm/vllm_worker.py [hunk 2]
```diff
@@ -616,7 +662,10 @@ def _build_sampling_params(
             top_k=top_k_val,
             max_tokens=max_tokens,
             logprobs=0,
-            stop_token_ids=self.cfg["stop_token_ids"],
+            stop_token_ids=self.cfg["stop_token_ids"]
+            if not self.cfg["ignore_eos"]
+            else [],
```
Comment (yuki-97) [inline nemo_rl/models/generation/vllm/vllm_worker.py]: wdyt moving this logic to `configure_generation_config` in `nemo_rl/models/generation/__init__.py`? so that we don't need to write it everywhere.

File: nemo_rl/models/generation/vllm/vllm_worker.py [hunk 3]
```diff
@@ -145,6 +145,32 @@ def __init__(
...
+                pass
+            else:
+                raise ValueError(
+                    f"Invalid output_len_or_output_len_generator: {output_len_or_output_len_generator}"
+                )
+            self.cfg["output_len_or_output_len_generator"] = (
+                output_len_or_output_len_generator
+            )
+        else:
+            self.cfg["output_len_or_output_len_generator"] = None
```
Comment (yuki-97) [inline nemo_rl/models/generation/vllm/vllm_worker.py]: 1. can we even make it a generator when `output_len_or_output_len_generator` is `int` (same for the input one)? then we can just have a `self.output_len_generator` (callable for benchmark and None for common case).
2. wdyt moving the whole logic to `get_sequence_length_generator`? since we use it in many places.

File: nemo_rl/models/generation/vllm/vllm_worker.py [hunk 4]
```diff
@@ -594,6 +620,26 @@ def _merge_stop_strings(self, batch_stop_strings):
 
         return list(stop_set) if stop_set else None
 
+    def _resolve_max_tokens(
```
Comment (yuki-97) [inline nemo_rl/models/generation/vllm/vllm_worker.py:610]: what about adding a comment int this function saying it's just used for benchmark?

File: tests/functional/L1_Functional_Tests_GPU.sh [hunk 5]
```diff
@@ -46,11 +46,13 @@ run_test      uv run --no-sync bash ./tests/functional/dpo_automodel_lora.sh
 run_test fast uv run --no-sync bash ./tests/functional/dpo_megatron_lora.sh
 run_test      uv run --no-sync bash ./tests/functional/dpo_megatron.sh
 run_test      uv run --no-sync bash ./tests/functional/eval.sh
+run_test      uv run --no-sync bash ./tests/functional/eval_random_dataset.sh
```
Comment (yuki-97) [inline tests/functional/L1_Functional_Tests_GPU.sh]: let's mark them with fast, so that we can use `Lfast` for test instead of `L1`, which will be faster a lot.

File: examples/run_eval.py [hunk 6]
```diff
@@ -100,6 +147,13 @@ def main():
     # Init ray
     init_ray()
 
+    is_random_dataset = config["data"].get("dataset_name") == "random"
+
+    # For the random path we don't need real model weights — vLLM can use dummy
+    # weights since outputs are not validated against any ground truth.
+    if is_random_dataset:
+        config["generation"]["vllm_cfg"]["load_format"] = "dummy"
```
Comment (yuki-97) [inline examples/run_eval.py]: need to move to under line 160, otherwise `load_format` will be override in `configure_generation_config`.

**Review by yuki-97 — COMMENTED** [review]
thanks @guyueh1 , overall lgtm, left some comments.

File: examples/run_eval.py (hunk 1 above)
Comment (guyueh1) [reply examples/run_eval.py]: agree, addressed in 35145d53

File: nemo_rl/models/generation/vllm/vllm_worker.py (hunk 2 above)
Comment (guyueh1) [reply nemo_rl/models/generation/vllm/vllm_worker.py]: agree, updated

File: nemo_rl/models/generation/vllm/vllm_worker.py (hunk 4 above)
Comment (guyueh1) [reply nemo_rl/models/generation/vllm/vllm_worker.py:610]: added docstring

File: tests/functional/L1_Functional_Tests_GPU.sh (hunk 5 above)
Comment (guyueh1) [reply tests/functional/L1_Functional_Tests_GPU.sh]: added

File: examples/run_eval.py (hunk 6 above)
Comment (guyueh1) [reply examples/run_eval.py]: done

File: nemo_rl/models/generation/vllm/vllm_worker.py (hunk 3 above)
Comment (guyueh1) [reply nemo_rl/models/generation/vllm/vllm_worker.py]: after thinking about this, i did a refactor; I breakdown the functionality of previous  `output_len_or_output_len_generator` to two variables:
* `output_len_or_output_distribution`: the config interface, allowed values are int (constant) or dict (parameters to a distribution)
* `output_len_generator`: the actual function that is used to sample output length
* function `get_sequence_length_generator` will parse `output_len_or_output_distribution` and return a callable `output_len_generator`
Same refactor is done for input len, wdyt

File: examples/run_eval.py [hunk 7]
```diff
@@ -39,17 +49,20 @@ def parse_args():
...
 
 
 def setup_data(tokenizer, data_config, env_configs):
     print("Setting up data...")
 
+    # Synthetic random-data path for benchmarking (no real dataset / env).
+    if data_config.get("dataset_name") == "random":
+        dataset, _, task_to_env, _ = setup_random_data(tokenizer, data_config)
+        env = task_to_env["random"]
+        return dataset, env, tokenizer
```
Comment (yuki-97) [inline examples/run_eval.py]: feels better to move this part to main() like in grpo.py, wdyt?

File: tests/functional/L1_Functional_Tests_GPU.sh (hunk 5 above)
Comment (yuki-97) [reply tests/functional/L1_Functional_Tests_GPU.sh]: let's mark fast also for `./tests/functional/grpo_random_dataset.sh`, then you can take a try of `CI:Lfast`, we usually use it for general PRs now and it's really quick a lot :)

File: nemo_rl/models/generation/interfaces.py [hunk 8]
```diff
@@ -116,7 +116,22 @@ class ColocationConfig(TypedDict):
...
+        backend: The backend to use for generation.
+        max_new_tokens: The maximum number of tokens to generate.
+        temperature: The temperature for sampling.
+        top_p: The top-p sampling parameter.
+        top_k: The top-k sampling parameter.
+        model_name: The name of the model.
+        stop_token_ids: The list of token IDs to stop generation.
+        stop_strings: The list of strings to stop generation.
+        ignore_eos: Whether to ignore the EOS token. This is only used for performance benchmarking purposes.
+        output_len_or_output_len_distribution: An integer or a dictionary with keys 'mean' and 'std' for the normal distribution that samples the output length. This is only used for performance benchmarking purposes.
```
Comment (yuki-97) [inline nemo_rl/models/generation/interfaces.py]: typo
```suggestion
        output_len_or_output_distribution: An integer or a dictionary with keys 'mean' and 'std' for the normal distribution that samples the output length. This is only used for performance benchmarking purposes.
```

File: nemo_rl/models/generation/vllm/vllm_worker.py [hunk 9]
```diff
@@ -594,6 +604,40 @@ def _merge_stop_strings(self, batch_stop_strings):
...
```
Comment (yuki-97) [inline nemo_rl/models/generation/vllm/vllm_worker.py:610]: do we also need to use this function in `generate_text_async`?

File: nemo_rl/utils/sequence_length_generator.py [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+        A callable that when invoked returns a sampled sequence length (int >= 1).
+    """
+    if constant_length_or_length_distribution is None:
+        return lambda _: None
+    if isinstance(constant_length_or_length_distribution, int):
+        return lambda _: max(1, constant_length_or_length_distribution)
+    if isinstance(constant_length_or_length_distribution, dict):
+        mean = constant_length_or_length_distribution["mean"]
+        std = constant_length_or_length_distribution["std"]
+        def sample_length(sample_idx: int | None = None) -> int:
```
Comment (yuki-97) [inline nemo_rl/utils/sequence_length_generator.py:39]: `sample_idx` looks unused. do we need to use it or can we drop it?

File: nemo_rl/models/generation/vllm/vllm_worker.py (hunk 3 above)
Comment (yuki-97) [reply nemo_rl/models/generation/vllm/vllm_worker.py]: thanks, looks good! and wdyt the change in the below? put it in the config feels not so good.

```
        output_len_generator = get_sequence_length_generator(output_len_or_output_distribution)
        self.cfg["output_len_generator"] = output_len_generator
```

to 

```
        self.output_len_generator = get_sequence_length_generator(output_len_or_output_distribution)
```

File: examples/run_eval.py (hunk 7 above)
Comment (guyueh1) [reply examples/run_eval.py]: done

File: nemo_rl/models/generation/vllm/vllm_worker.py (hunk 3 above)
Comment (guyueh1) [reply nemo_rl/models/generation/vllm/vllm_worker.py]: yeah, i think that make sense, done

File: tests/functional/L1_Functional_Tests_GPU.sh (hunk 5 above)
Comment (guyueh1) [reply tests/functional/L1_Functional_Tests_GPU.sh]: done

File: nemo_rl/models/generation/interfaces.py (hunk 8 above)
Comment (guyueh1) [reply nemo_rl/models/generation/interfaces.py]: changed

File: nemo_rl/utils/sequence_length_generator.py (hunk 10 above)
Comment (guyueh1) [reply nemo_rl/utils/sequence_length_generator.py:39]: i kept it because I want to make the generator to take one argument, though it's not used here; suppose in the future we want to extend it to some more complex callable based on sample_idx, we can .

Comment (guyueh1) [reply nemo_rl/utils/sequence_length_generator.py:39]: it can be omitted when invoking this function because it has a default value

File: nemo_rl/models/generation/vllm/vllm_worker.py (hunk 9 above)
Comment (guyueh1) [reply nemo_rl/models/generation/vllm/vllm_worker.py:610]: it should.. though i don't think in practice there's a scenario we use it, since when we want to do benchmarking we just use the generate_async . I added it anyway

File: nemo_rl/models/generation/vllm/vllm_worker.py [hunk 11]
```diff
@@ -145,6 +145,18 @@ def __init__(
...
+            get_sequence_length_generator,
+        )
+
+        output_len_or_output_distribution = self.cfg.get(
+            "output_len_or_output_distribution", None
+        )
+        output_len_generator = get_sequence_length_generator(
+            output_len_or_output_distribution
+        )
+        self.output_len_generator = output_len_generator
```
Comment (yuki-97) [inline nemo_rl/models/generation/vllm/vllm_worker.py]: nit
```suggestion
        self.output_len_generator = get_sequence_length_generator(
            output_len_or_output_distribution
        )
```

**Review by yuki-97 — DISMISSED** [review]
LGTM, thanks @guyueh1

**Comment by guyueh1** [comment]
CI is green, @terrykong  to take a look

File: nemo_rl/evals/eval.py [hunk 12]
```diff
@@ -152,6 +153,7 @@ def setup(
     # check backend
     backend = generation_config["backend"]
     assert backend == "vllm", "Only vLLM backend is supported for evaluation"
+    generation_config = cast(VllmConfig, generation_config)
```
Comment (guyueh1) [inline nemo_rl/evals/eval.py]: I think this was left during merge conflict resolution; will remove later

**Review by yuki-97 — APPROVED** [review]

**Comment by guyueh1** [comment]
@terrykong please review

**Comment by guyueh1** [comment]
i will resolve the merge conflict and do final rebase & testing after @terrykong  reviews

File: nemo_rl/evals/eval.py [hunk 13]
```diff
@@ -411,7 +413,8 @@ async def _run_env_eval_impl(
             )
         else:
             raise ValueError(f"Invalid metric: {metric}")
-
+        step_time = time.time() - start_time
+        print(f"Step time: {step_time:.2f}s")
```
Comment (terrykong) [inline nemo_rl/evals/eval.py:417]: [`nemo_rl/evals/eval.py:324,416-417`](https://github.com/NVIDIA-NeMo/RL/blob/735b32beac58cf7d245b89c34aefa8a3b7c52251/nemo_rl/evals/eval.py#L324)

Leftover debug/profiling code: unconditional `start_time = time.time()` (L324) and `print(f"Step time: ...")` (L417) run on **every** eval step, not just the random/benchmark path. This was [acknowledged](https://github.com/NVIDIA-NeMo/RL/pull/1453#discussion_r3267732981) as leftover merge-conflict debug code that should be removed. Consider removing it or gating behind a logger/debug flag.

File: nemo_rl/data/processors.py [hunk 14]
```diff
@@ -707,6 +707,49 @@ def nemo_gym_data_processor(
...
+        raise ValueError(
+            "input_len_generator must be provided if you are using random_input_len_processor"
+        )
+
+    input_len = input_len_generator(idx)
+    assert input_len is not None, (
+        "input_len_or_input_distribution must be provided when data.dataset_name == 'random'"
+    )
+    if max_seq_length is not None:
+        assert input_len <= max_seq_length
```
Comment (terrykong) [inline nemo_rl/data/processors.py:729]: [`nemo_rl/data/processors.py:729`](https://github.com/NVIDIA-NeMo/RL/blob/735b32beac58cf7d245b89c34aefa8a3b7c52251/nemo_rl/data/processors.py#L729)

With a normal-distribution `input_len_or_input_distribution`, a sampled length above `max_input_seq_length` aborts the whole run mid-iteration via bare `assert`. Two concerns:
1. `assert` is stripped under `python -O`, so the guard silently disappears
2. The output side (`_resolve_max_tokens`) clamps with `min()` — consider clamping here too, or raising a `ValueError` with context

```suggestion
    if max_seq_length is not None and input_len > max_seq_length:
        input_len = max_seq_length
```

File: nemo_rl/models/generation/vllm/vllm_worker.py [hunk 15]
```diff
@@ -871,8 +915,9 @@ def generate_text(
             temperature=self.cfg["temperature"] if not greedy else 0,
             top_p=self.cfg["top_p"],
             top_k=top_k if not greedy else 1,
-            max_tokens=self.cfg["max_new_tokens"],
+            max_tokens=self._resolve_max_tokens(),
```
Comment (terrykong) [inline nemo_rl/models/generation/vllm/vllm_worker.py:918]: [`nemo_rl/models/generation/vllm/vllm_worker.py:918`](https://github.com/NVIDIA-NeMo/RL/blob/735b32beac58cf7d245b89c34aefa8a3b7c52251/nemo_rl/models/generation/vllm/vllm_worker.py#L918)

Sync path calls `_resolve_max_tokens()` with `sample_idx=None` → one sampled output length per batch, not per sample. Per-sample OSL only works in the async path. GRPO defaults to `async_engine: false` (sync). Worth documenting this limitation for users of `{{mean, std}}` output distributions.

File: nemo_rl/data/utils.py [hunk 16]
```diff
@@ -390,3 +396,75 @@ def setup_preference_data(
...
+
+    dataset = AllTaskProcessedDataset(
+        base_dataset.formatted_ds["train"],
+        tokenizer,
+        random_task_spec,
+        task_data_processors,
+        max_seq_length=data_config["max_input_seq_length"],
+    )
+
+    val_dataset: Optional[AllTaskProcessedDataset] = None
```
Comment (terrykong) [inline nemo_rl/data/utils.py:466]: [`nemo_rl/data/utils.py:466`](https://github.com/NVIDIA-NeMo/RL/blob/735b32beac58cf7d245b89c34aefa8a3b7c52251/nemo_rl/data/utils.py#L466)

`setup_random_data` always returns `val_dataset=None`. If GRPO validation is enabled (`val_at_start`/`val_period>0`), it will fail. The functional test disables both. Could you add a guard with a clear error message, or document this limitation?

File: nemo_rl/data/utils.py [hunk 17]
```diff
@@ -390,3 +396,75 @@ def setup_preference_data(
...
+    )
+
+    base_dataset: Any = RandomDataset(
+        input_len_generator,
+        num_samples=data_config["num_samples"],
+    )
+
+    # data processor — defaultdict so any task name in the synthetic data falls
+    # back to the random processor
+    task_data_processors: dict[str, tuple[TaskDataSpec, TaskDataProcessFnCallable]] = (
```
Comment (terrykong) [inline nemo_rl/data/utils.py:444]: [`nemo_rl/data/utils.py:444-447`](https://github.com/NVIDIA-NeMo/RL/blob/735b32beac58cf7d245b89c34aefa8a3b7c52251/nemo_rl/data/utils.py#L444)

Nit: this `defaultdict` fallback is misleading — `AllTaskProcessedDataset.__getitem__` uses `assert task_name in self.task_data_processors`, and Python's `in` operator does NOT trigger a defaultdict's `__missing__`. The fallback never fires. Works only because synthetic data always emits `task_name="random"`. A plain `dict` would be clearer. Same for `task_to_env` at L468.

File: nemo_rl/models/generation/interfaces.py [hunk 18]
```diff
@@ -126,6 +141,8 @@ class GenerationConfig(TypedDict):
     model_name: NotRequired[str]  # Not Required b/c GRPO writes this
     stop_token_ids: list[int] | None
     stop_strings: list[str] | None
+    ignore_eos: bool
```
Comment (terrykong) [inline nemo_rl/models/generation/interfaces.py:144]: [`nemo_rl/models/generation/interfaces.py:144`](https://github.com/NVIDIA-NeMo/RL/blob/735b32beac58cf7d245b89c34aefa8a3b7c52251/nemo_rl/models/generation/interfaces.py#L144)

Making `ignore_eos` a required field is guideline-compliant and all in-repo configs are updated. However, this is a backward-incompatible change for external user YAMLs predating this PR — they'll get `KeyError`. A release note or migration note in the PR description would be helpful.

File: nemo_rl/models/generation/__init__.py [hunk 19]
```diff
@@ -69,4 +69,8 @@ def configure_generation_config(
             else:
                 config["vllm_cfg"]["skip_tokenizer_init"] = True
 
+        # set stop_token_ids to empty list if ignore_eos is True
+        if config["ignore_eos"]:
```
Comment (terrykong) [inline nemo_rl/models/generation/__init__.py:73]: [`nemo_rl/models/generation/__init__.py:73`](https://github.com/NVIDIA-NeMo/RL/blob/735b32beac58cf7d245b89c34aefa8a3b7c52251/nemo_rl/models/generation/__init__.py#L73)

**[BUG]** This unconditional access of `config["ignore_eos"]` will `KeyError` in two test files that build vllm-backend configs without the new required field:
- [`test_vllm_quant_backend.py:50-81`](https://github.com/NVIDIA-NeMo/RL/blob/735b32beac58cf7d245b89c34aefa8a3b7c52251/tests/unit/models/generation/test_vllm_quant_backend.py#L50-L81) — breaks on GPU+modelopt CI
- [`test_sglang_generation.py:37`](https://github.com/NVIDIA-NeMo/RL/blob/735b32beac58cf7d245b89c34aefa8a3b7c52251/tests/unit/models/generation/test_sglang_generation.py#L37) — runtime-safe (SGLang doesn't read it), but pyrefly/type concern

Fix: add `"ignore_eos": False` to both test configs.

File: tests/unit/models/generation/test_vllm_generation.py [hunk 20]
```diff
@@ -85,6 +86,32 @@
     "vllm_kwargs": {},
 }
 
+
+def test_base_vllm_worker_converts_output_len_generator_dict():
```
Comment (terrykong) [inline tests/unit/models/generation/test_vllm_generation.py:90]: **[TEST]** Several new functions lack direct unit tests (all CPU-testable):
- [`_resolve_max_tokens()`](https://github.com/NVIDIA-NeMo/RL/blob/735b32beac58cf7d245b89c34aefa8a3b7c52251/nemo_rl/models/generation/vllm/vllm_worker.py#L610) — no test despite real branch logic (None, <max, >max paths). Top priority.
- [`get_sequence_length_generator`](https://github.com/NVIDIA-NeMo/RL/blob/735b32beac58cf7d245b89c34aefa8a3b7c52251/nemo_rl/utils/sequence_length_generator.py#L19) — only dict branch tested; int/None/callable untested
- [`random_input_len_processor`](https://github.com/NVIDIA-NeMo/RL/blob/735b32beac58cf7d245b89c34aefa8a3b7c52251/nemo_rl/data/processors.py#L710) — 3 error paths untested (generator=None, generator→None, missing vocab_size)
- [`setup_random_data()`](https://github.com/NVIDIA-NeMo/RL/blob/735b32beac58cf7d245b89c34aefa8a3b7c52251/nemo_rl/data/utils.py#L401) — no unit test; early validation branches are CPU-testable
- [`RandomDataset`](https://github.com/NVIDIA-NeMo/RL/blob/735b32beac58cf7d245b89c34aefa8a3b7c52251/nemo_rl/data/datasets/random_dataset.py#L23) — edge cases (num_samples=0, negative idx)

File: nemo_rl/utils/sequence_length_generator.py [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+        return lambda _: None
+    if isinstance(constant_length_or_length_distribution, int):
+        sequence_length = max(1, constant_length_or_length_distribution)
+        return lambda _: sequence_length
+    if isinstance(constant_length_or_length_distribution, dict):
+        mean = constant_length_or_length_distribution["mean"]
+        std = constant_length_or_length_distribution["std"]
+
+        def sample_length(sample_idx: int | None = None) -> int:
+            length = int(np.round(np.random.normal(mean, std)))
```
Comment (terrykong) [inline nemo_rl/utils/sequence_length_generator.py:40]: [`sequence_length_generator.py:40`](https://github.com/NVIDIA-NeMo/RL/blob/735b32beac58cf7d245b89c34aefa8a3b7c52251/nemo_rl/utils/sequence_length_generator.py#L40)

**[LOW]** Neither `np.random.normal` here nor `torch.randint` ([processors.py:739](https://github.com/NVIDIA-NeMo/RL/blob/735b32beac58cf7d245b89c34aefa8a3b7c52251/nemo_rl/data/processors.py#L739)) is seeded. With `num_workers>0`, forked NumPy RNG state can yield identical sequences across workers. Low impact for benchmarking but reproducibility is usually desirable.

File: nemo_rl/models/generation/vllm/vllm_worker.py [hunk 22]
```diff
@@ -596,6 +607,38 @@ def _merge_stop_strings(self, batch_stop_strings):
 
         return list(stop_set) if stop_set else None
 
+    def _resolve_max_tokens(
+        self, max_new_tokens: Optional[int] = None, sample_idx: int | None = None
```
Comment (terrykong) [inline nemo_rl/models/generation/vllm/vllm_worker.py:611]: @guyueh1 not sure where to add such comment, but do you think we should document somewhere that the way in which this method is called is different between sync vs async path?

IIUC, the sync path picks a max length per batch and all N samples get that output length, but the async path looks like it's set per sample, so if you have a long ISL, your output tokens are clamped more aggressively (so benchmarking just changing the async_engine=true|false seems like it'd have this nuance)

File: nemo_rl/models/generation/interfaces.py (hunk 18 above)
Comment (terrykong) [reply nemo_rl/models/generation/interfaces.py:144]: maybe a a comment above this that the before/default behavior is `False`, just in case someone wonders what they should be setting this to when they rebase and their configs don't have this key

**Review by terrykong — COMMENTED** [review]
## Review: PR #1453 — Random Dataset with Specified Input and Output Sequence Length

> ⚠️ **Merge conflict**: PR is `CONFLICTING` — please rebase on `main` and resolve conflicts before further review.

All findings are posted as inline comments on the relevant files/lines.

_Generated by Claude Code_
